File: src/eqn/ast/EqnAstNodePower.java
package eqn.ast;

import eqn.parser.exception.EqnException;

public class EqnAstNodePower extends EqnAstNodeBinaryOperation {

    public EqnAstNodePower(EqnAstNode leftOperand, EqnAstNode rightOperand) throws EqnException {
        super("^", leftOperand, rightOperand, PrecedenceType.Power);
    }

    @Override
    public double evaluate() throws EqnException {
        return Math.pow(left().evaluate(), right().evaluate());
    }

    @Override
    public EqnAstNode simplify() throws EqnException {
        simplifyChildren();

        if (right().type == Type.Constant) {
            if (Double.parseDouble(right().value) == 0.0D) {
                return new EqnAstNodeDouble(1.0D);
            }
            if (Double.parseDouble(right().value) == 1.0D) {
                return left();
            }
        }
        if (left().type == Type.Constant) {
            if (Double.parseDouble(left().value) == 1.0D) {
                return new EqnAstNodeDouble(1.0D);
            }
        }
        return super.simplify();
    }
}

File: src/eqn/ast/EqnAstNodePower.kt
package eqn.ast

import eqn.ast.base.EqnAstNode
import eqn.ast.base.EqnAstNodeBinary
import eqn.parser.exception.EqnException
import kotlin.math.pow

class EqnAstNodePower(leftOperand: EqnAstNode, rightOperand: EqnAstNode) : EqnAstNodeBinary("^", Type.Power, PrecedenceType.Power, leftOperand, rightOperand) {

    @Throws(EqnException::class)
    override fun evaluate(arguments: Map<String, Double>?): Double {
        return left.evaluate(arguments).pow(right.evaluate(arguments))
    }

    @Throws(EqnException::class)
    override fun simplify(arguments: Map<String, Double>?): EqnAstNode {
        simplifyChildren(arguments)
        if (right.type == Type.Constant) {
            if (right.value.toDouble() == 0.0) {
                return EqnAstNodeDouble(1.0)
            }
            if (right.value.toDouble() == 1.0) {
                return left
            }
        }
        if (left.type == Type.Constant) {
            if (left.value.toDouble() == 1.0) {
                return EqnAstNodeDouble(1.0)
            }
        }
        return super.simplify(arguments)
    }
}